redhat-appstudio/application-service
# [DEVHAS-21] Add webhook scaffold & logic for update

Signed-off-by: Maysun J Faisal <maysunaneek@gmail.com>

Fixes https://issues.redhat.com/browse/DEVHAS-21 and https://issues.redhat.com/browse/DEVHAS-20

## Summary

This PR adds the webhook validation for `Application` & `Component` update. The validating webhook blocks the client from submitting CR updates to fields that are immutable acc to [doc](https://docs.google.com/document/d/1dfkTky3qFv6wTXdMX9Ag-bYFEt3OfBgRZOceYNSwyEU/edit?pli=1#heading=h.6b6ftdufj5y3).

It also add create validation checks for the `Component` CR, to make sure at least one source either git or image is specified and if a git source is specified, the URL is a valid URL.

The `Application` spec has been updated to make sure the `displayName` is required, the `Component` spec has been updated to add regex and char limit for `application` & `componentName`

The Makefile also has new commands to install and uninstall certificate manager kube resources to add the CR for self-signed certificates as required by [admission webhooks](https://sdk.operatorframework.io/docs/building-operators/golang/webhook/) since they intercept the API request.

The PR also adds unit and integration tests. The webhook integration tests  does not check reconciler logic as the k8sClient generated by the scaffold does not add the Reconcilers to the manager for the testEnv.

## How to test it?

Create an `Application` & `Component` as you would normally do. Go either to the `Application` or the `Component` yaml that was used and update any immutable field and apply. You should see that the server blocks the update.

For example, I updated the component name in the `Component` yaml from `backend` to `backends`:
```
$ k apply -f config/samples/component/component-basic.yaml 
Error from server (component name cannot be updated to backends): error when applying patch:
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"appstudio.redhat.com/v1alpha1\",\"kind\":\"Component\",\"metadata\":{\"annotations\":{},\"name\":\"component-sample\",\"namespace\":\"application-service-system\"},\"spec\":{\"application\":\"application-sample\",\"componentName\":\"backends\",\"env\":[{\"name\":\"FOO\",\"value\":\"foo1\"},{\"name\":\"BAR\",\"value\":\"bar1\"}],\"replicas\":1,\"resources\":{\"limits\":{\"cpu\":\"2\",\"ephemeral-storage\":\"500Mi\",\"memory\":\"500Mi\",\"storage\":\"400Mi\"},\"requests\":{\"cpu\":\"700m\",\"ephemeral-storage\":\"400Mi\",\"memory\":\"400Mi\",\"storage\":\"200Mi\"}},\"route\":\"route111\",\"source\":{\"git\":{\"url\":\"https://github.com/devfile-samples/devfile-sample-java-springboot-basic\"}},\"targetPort\":1111}}\n"}},"spec":{"componentName":"backends"}}
to:
Resource: "appstudio.redhat.com/v1alpha1, Resource=components", GroupVersionKind: "appstudio.redhat.com/v1alpha1, Kind=Component"
Name: "component-sample", Namespace: "application-service-system"
for: "config/samples/component/component-basic.yaml": admission webhook "vcomponent.kb.io" denied the request: component name cannot be updated to backends
```

## Codecov

I added a new codecov config file to ignore the generated deepcopy file from the operator-sdk. Since the `v1alpha1` dir now has *_test.go files, this dir is now included in the codecov report. This PR so far increases the total coverage of the project by 3%

<img width="1292" alt="Screen Shot 2022-01-07 at 3 09 53 PM" src="https://user-images.githubusercontent.com/31771087/148601591-3f82e6bf-1982-4cb7-b8d8-8ebf96c4d8b4.png">
